File: code-sample-java/code-sample-java-algorithms-and-data-structures/src/main/java/codesample/leetcode/medium/_22_GenerateParentheses.java
package codesample.leetcode.medium;

import java.util.ArrayList;
import java.util.List;

/**
 * 22. Generate Parentheses — https://leetcode.com/problems/generate-parentheses/
 * <p>
 * Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
 * Example 1:
 * <p>
 * Input: n = 3
 * Output: ["((()))","(()())","(())()","()(())","()()()"]
 * Example 2:
 * <p>
 * Input: n = 1
 * Output: ["()"]
 */
public class _22_GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        int opened = 1;
        List<String> good = new ArrayList<>();

        allGoodParenthesis(opened, n * 2 - 1, "(", good);

        return good;
    }


    public void allGoodParenthesis(int opened, int parenthesisLeft, String current, List<String> result) {
        if (parenthesisLeft == 0) {
            result.add(current);
            return;
        }

        if (opened > 0 && parenthesisLeft > opened) { // some opened, but can open more
            allGoodParenthesis(opened + 1, parenthesisLeft - 1, current + "(", result);
        }
        if (opened == 0 && parenthesisLeft >= 2) { // none opened, but can open + close
            allGoodParenthesis(opened + 1, parenthesisLeft - 1, current + "(", result);
        }
        if (opened > 0) { // close opened
            allGoodParenthesis(opened - 1, parenthesisLeft - 1, current + ")", result);
        }
    }
}

File: code-sample-kotlin/algorithms/src/main/kotlin/com/codesample/leetcode/medium/22_generateParentheses.kt
package com.codesample.leetcode.medium

import java.lang.StringBuilder

/** 22. Generate Parentheses https://leetcode.com/problems/generate-parentheses/
 *
 * Given n, output every correct parenthesis sequence with size 2*n. Input should be sorted lexicographically
 * e.g:
 * n = 1: ()
 * n = 2: (()),()()
 * n = 3: ((())),(()()),(())(),()(()),()()()
 * */

fun generateParenthesis(n: Int): List<String> {

    val opened = 1
    val good = mutableListOf<String>();

    fun allGoodParenthesis(opened: Int, parenthesisLeft: Int, current: String) {
        if (parenthesisLeft == 0) {
            good.add(current)
            return
        }
        if (opened > 0 && parenthesisLeft > opened) { // some opened, but can open more
            allGoodParenthesis(opened + 1, parenthesisLeft - 1, "$current(")
        }
        if (opened == 0 && parenthesisLeft >= 2) { // none opened, but can open + close
            allGoodParenthesis(opened + 1, parenthesisLeft - 1, "$current(")
        }
        if (opened > 0) { // close opened
            allGoodParenthesis(opened - 1, parenthesisLeft - 1, "$current)")
        }

    }

    allGoodParenthesis(opened, n*2 - 1, "(")
    return good.map { "\"$it\"" }
}

fun main() {
    println(generateParenthesis(1)) // ()
    println(generateParenthesis(2)) // (()),()()
    println(generateParenthesis(3)) // ((())),(()()),(())(),()(()),()()()
}
